File: py/leetcode-maxproddiff.py
class Solution:
    def maxProductDifference(self, nums: List[int]) -> int:
        maxi = 0
        smaxi = 1 if nums[0] > nums[1] else 0

        lowi = 0
        slowi = 1 if nums[0] < nums[1] else 0

        for i in range(len(nums)):
            if nums[i] > nums[maxi]:
                smaxi = maxi
                maxi = i
            elif nums[i] > nums[smaxi] and i != maxi:
                smaxi = i
            if nums[i] < nums[lowi]:
                slowi = lowi
                lowi = i
            elif nums[i] < nums[slowi] and i != lowi:
                slowi = i

        print(f"maxi: {maxi} smaxi: {smaxi}")
        print(f"lowi: {lowi} slowi: {slowi}")

        return (nums[maxi] * nums[smaxi]) - (nums[lowi] * nums[slowi])

File: rs/leetcode-maxproddiff.rs
impl Solution {
    pub fn max_product_difference(nums: Vec<i32>) -> i32 {
        let mut maxi = 0;
        let mut smaxi = if nums[0] > nums[1] { 1 } else { 0 };

        let mut lowi = 0;
        let mut slowi = if nums[0] < nums[1] { 1 } else { 0 };

        for i in 0..nums.len() {
            if nums[i] > nums[maxi] {
                smaxi = maxi;
                maxi = i;
            }
            else if nums[i] > nums[smaxi] && i != maxi {
                smaxi = i;
            }
            if nums[i] < nums[lowi] {
                slowi = lowi;
                lowi = i;
            }
            else if nums[i] < nums[slowi] && i != lowi {
                slowi = i;
            }
        }

        (nums[maxi] * nums[smaxi]) - (nums[lowi] * nums[slowi])
    }
}